ntoskrnl/io/pnpmgr/pnpnotify.c

```diff
@@ -19,210 +19,274 @@
-                NotificationInfos->Size = sizeof(TARGET_DEVICE_REMOVAL_NOTIFICATION);
-                RtlCopyMemory(&NotificationInfos->Event, Event, sizeof(GUID));
-            }
-            else
-            {
-                PTARGET_DEVICE_CUSTOM_NOTIFICATION NotificationInfos;
-                NotificationStructure = NotificationInfos = ExAllocatePoolWithTag(
-                    PagedPool,
-                    sizeof(TARGET_DEVICE_CUSTOM_NOTIFICATION),
-                    TAG_PNP_NOTIFY);
-                if (!NotificationInfos)
-                {
-                    KeReleaseGuardedMutex(&PnpNotifyListLock);
-                    return;
-                }
-                RtlCopyMemory(NotificationInfos, EventCategoryData1, sizeof(TARGET_DEVICE_CUSTOM_NOTIFICATION));
-            }
-            break;
-        }
-        default:
-        {
-            DPRINT1("IopNotifyPlugPlayNotification(): unknown EventCategory 0x%x UNIMPLEMENTED\n", EventCategory);
-            KeReleaseGuardedMutex(&PnpNotifyListLock);
-            return;
-        }
+        return;
     }
 
-    /* Loop through procedures registred in PnpNotifyListHead
-     * list to find those that meet some criteria.
-     */
-    ListEntry = PnpNotifyListHead.Flink;
-    while (ListEntry != &PnpNotifyListHead)
+    // release the lock because the notification routine has to be called without any
+    // limitations regarding APCs
+    KeReleaseGuardedMutex(Lock);
+
+    // for verifying the driver behaviour
+#if DBG
+    KIRQL oldIrql = KeGetCurrentIrql();
+    ULONG oldApcDisable = KeGetCurrentThread()->CombinedApcDisable;
+#endif
+
+    Entry->PnpNotificationProc(NotificationStructure, Entry->Context);
+
+    ASSERT(oldIrql == KeGetCurrentIrql() &&
+           oldApcDisable == KeGetCurrentThread()->CombinedApcDisable);
+    KeAcquireGuardedMutex(Lock);
+
+    ExFreePoolWithTag(NotificationStructure, TAG_PNP_NOTIFY);
+}
+
+/**
+ * @brief      Delivers the event to all drivers subscribed to
+ *             EventCategoryDeviceInterfaceChange
+ *
+ * @param[in]  Event               The PnP event GUID
+ * @param[in]  InterfaceClassGuid  The GUID of an interface class
+ * @param[in]  SymbolicLinkName    Pointer to a string identifying the device interface name
+ */
+CODE_SEG("PAGE")
+VOID
+PiNotifyDeviceInterfaceChange(
+    _In_ LPCGUID Event,
+    _In_ LPCGUID InterfaceClassGuid,
+    _In_ PUNICODE_STRING SymbolicLinkName)
+{
+    PDEVICE_INTERFACE_CHANGE_NOTIFICATION notifyStruct;
+    notifyStruct = ExAllocatePoolWithTag(PagedPool, sizeof(*notifyStruct), TAG_PNP_NOTIFY);
+    if (!notifyStruct)
     {
-        ChangeEntry = CONTAINING_RECORD(ListEntry, PNP_NOTIFY_ENTRY, PnpNotifyList);
-        CallCurrentEntry = FALSE;
+        return;
+    }
 
-        if (ChangeEntry->EventCategory != EventCategory)
-        {
-            ListEntry = ListEntry->Flink;
-            continue;
-        }
+    *notifyStruct = (DEVICE_INTERFACE_CHANGE_NOTIFICATION) {
```

Comment:
Not standard-conforming.

Reply:
More importantly, I don't think this provides a readability benefit over assigning the members individually, and in fact makes the code two lines longer.

Reply:
It also zeroes all non-assigned members, so you're saving `RtlZeroMemory` call here